### PR TITLE
external: pre-alloc mem for writer, use offset to avoid golang slice

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #47572

Problem Summary:

### What is changed and how it works?

pre-alloc memory for writer, we align to block-size, default is 16MiB, for import-into data kv we use 32MiB
store each KV inside buf directly and we use a custom location to pointer to it, to avoid the overhead of golang slice, reduce from 3*24(golang slice size) to 16.
this way we can have less smaller object, and can use memory more efficiently, less GC

by using this way, we put limition on KV size, the max kv size is 16MiB for index, and 32MiB for data.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
in below test, we set those 2 params to avoid tidb mem tunner causes too many GC as it's known global sort will uses  a lot of mem
```sql
mysql> show variables like 'tidb_server_memory_limit_gc_trigger';
+-------------------------------------+-------+
| Variable_name                       | Value |
+-------------------------------------+-------+
| tidb_server_memory_limit_gc_trigger | 0.99  |
+-------------------------------------+-------+
1 row in set (0.04 sec)

mysql> show variables like 'tidb_server_memory_limit';
+--------------------------+-------+
| Variable_name            | Value |
+--------------------------+-------+
| tidb_server_memory_limit | 88%   |
+--------------------------+-------+
```

after this pr, on a case of 158.8GB, import takes 46m(after) vs 52m(before). the time takes on `encode&sort` is 12m2s(after) vs 18m45(before). after this pr, the time takes on `encode&sort` step is close to `local sort`, and the encode speed is more stable

before:
image

after
image

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by wjhuang2016 — APPROVED** [review]

File: br/pkg/lightning/backend/external/file.go [hunk 1]
```diff
@@ -46,51 +49,26 @@ func NewKeyValueStore(
...
-	n, err = s.dataWriter.Write(
-		s.ctx,
-		binary.BigEndian.AppendUint64(b[:0], uint64(len(value))),
-	)
-	if err != nil {
-		return err
-	}
-	kvLen += n
-	n, err = s.dataWriter.Write(s.ctx, value)
+func (s *KeyValueStore) AddData(val []byte) error {
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/file.go:57]: Just noticed no other package is using this function so we can make it private. Another reason is now the function need caller to build data in specific layout, which is not friendly as a public function.

File: br/pkg/lightning/backend/external/kv_buf.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+// allocation, all space in this buffer will be reused when reset.
+type preAllocKVBuf struct {
+	blocks      [][]byte
+	blockSize   int
+	curBlock    []byte
+	curBlockIdx int
+	curIdx      int
+}
+
+func newPreAllocKVBuf(memLimit uint64, blockSize int) *preAllocKVBuf {
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/kv_buf.go:32]: it looks similar to membuf.Buffer 🤔 not sure we can reuse it

File: br/pkg/lightning/backend/external/writer.go [hunk 3]
```diff
@@ -304,22 +319,36 @@ type Writer struct {
...
-	if w.batchSize >= w.memSizeLimit {
+	encodedKeyLen := keyAdapter.EncodedLen(idxKey, rowID)
+	length := encodedKeyLen + len(idxVal) + lengthBytes*2
+	blockIdx, dataBuf, off, allocated := w.kvBuffer.Alloc(length)
+	if !allocated {
 		if err := w.flushKVs(ctx, false); err != nil {
 			return err
 		}
+		blockIdx, dataBuf, off, allocated = w.kvBuffer.Alloc(length)
+		// we now don't support KV larger than blockSize
```
Comment (lance6716) [inline br/pkg/lightning/backend/external/writer.go:335]: seems we will panic? And why do we need to split `memLimit` into multiple blocks.

Comment (D3Hunter) [reply br/pkg/lightning/backend/external/writer.go:335]: why panic? the mem-limit might be several GB on add-index, it's too large, and hasn't test, but the performance should be similar.

Comment (lance6716) [reply br/pkg/lightning/backend/external/writer.go:335]: I changed an unit test of memLimit = 16, blockSize = 8, Alloc(9). Alloc found the current block is not enough to hold the data and moved to next block, but does not check the pre-allocated blocks have enough capacity

File: br/pkg/lightning/backend/external/kv_buf.go (hunk 2 above)
Comment (D3Hunter) [reply br/pkg/lightning/backend/external/kv_buf.go:32]: fields are similar, but not the code, will leave it for now

File: br/pkg/lightning/backend/external/file.go (hunk 1 above)
Comment (D3Hunter) [reply br/pkg/lightning/backend/external/file.go]: https://github.com/pingcap/tidb/pull/47793/commits/c8e33104edb36bdca262e713451c77a20dca0953

File: br/pkg/lightning/backend/external/writer.go (hunk 3 above)
Comment (D3Hunter) [reply br/pkg/lightning/backend/external/writer.go:335]: https://github.com/pingcap/tidb/pull/47793/commits/0d842cb3a6d1d2e573e0bfc7e4d9eee32d0fa50c

File: pkg/disttask/importinto/encode_and_sort_operator.go [hunk 4]
```diff
@@ -250,3 +257,10 @@ func getNumOfIndexGenKV(tblInfo *model.TableInfo) int {
 	}
 	return count
 }
+
+func getKVGroupBlockSize(group string) int {
```
Comment (lance6716) [inline pkg/disttask/importinto/encode_and_sort_operator.go:261]: I still worried about the block size is not enough. In lightning's CSV parser, we can tolerate ~120M KV size (however the only case user get such a big KV pair is that line terminator is not correct / source file is corrupted so the multiple lines are parsed as one line)

https://github.com/pingcap/tidb/blob/b0c01fa9a44e0abbbed9abca306aefd29e16bda2/br/pkg/lightning/mydump/csv_parser.go#L44

And if we don't split `memLimit` into multiple blocks, we will not meet this problem. I don't have strong oponion about this suggestion

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: pkg/disttask/importinto/encode_and_sort_operator.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/disttask/importinto/encode_and_sort_operator.go:261]: discussed with @Benjamin2037 last week, plan to add a restriction in doc about kv size when import using global sort

> a big KV pair is that line terminator is not correct / source file is corrupted

for this case, if multiple line is concated and parsed successfully, we can identified it easily

**Pull request merged**